src/Xrefcheck/Verify.hs

```diff
@@ -426,94 +455,107 @@ verifyReference
-      :: (Num a)
-      => Int
-      -> Progress a
-      -> Progress a
-    alterOverallProgress retryNumber
-      | retryNumber > 0 = id
-      | otherwise = incProgress
-
-    alterProgressErrors
-      :: (Num a)
-      => VerifyResult VerifyError
-      -> Int
-      -> Progress a
-      -> Progress a
-    alterProgressErrors res@(VerifyResult ves) retryNumber
-      | (ncMaxRetries cNetworking) == 0 =
-          if ok then id
-          else incProgressUnfixableErrors
-      | retryNumber == 0 =
-          if ok then id
-          else if fixable then incProgressFixableErrors
-          else incProgressUnfixableErrors
-      | retryNumber == (ncMaxRetries cNetworking) =
-          if ok then decProgressFixableErrors
-          else fixableToUnfixable
-      -- 0 < retryNumber < ncMaxRetries
-      | otherwise =
-          if ok then decProgressFixableErrors
-          else if fixable then id
-          else fixableToUnfixable
-      where
-        ok = verifyOk res
-        fixable = any isFixable ves
-
-    extractRetryAfterInfo :: VerifyResult VerifyError -> Maybe RetryAfter
-    extractRetryAfterInfo = \case
-      VerifyResult [ExternalHttpTooManyRequests retryAfter] -> retryAfter
+      case res of
+        -- Success
+        Right () -> do
+          let moveProgressOnSuccess = composeFuncList
+                [ unlessFunc (errorsOccured rc) incProgress
+                , whenFunc (errorsOccured rc) decProgressFixableErrors
+                ]
+          modifyProgressRef now Nothing moveProgressOnSuccess
+          pure res
+        Left err -> do
+          setOfReturned429 <- addDomainIf429 domainsReturned429Ref err
+          case decideWhetherToRetry setOfReturned429 rc err of
+            -- Unfixable
+            Nothing -> do
+              let moveProgressOnUnfixable = composeFuncList
+                    [ unlessFunc (errorsOccured rc) incProgress
+                    , if errorsOccured rc
+                      then fixableToUnfixable
+                      else incProgressUnfixableErrors
+                    ]
+              modifyProgressRef now Nothing moveProgressOnUnfixable
+              pure res
+            -- Fixable, retry
+            Just (mbCurrentRetryAfter, counterModifier) -> do
+              let toSeconds = \case
+                    Seconds s -> s
+                    -- Calculates the seconds left until @Retry-After@ date.
+                    -- Defaults to 0 if the date has already passed.
+                    Date date | utcTimeToTimeSecond date >= now -> utcTimeToTimeSecond date -:- now
+                    _ -> sec 0
+
+              let currentRetryAfter = fromMaybe (ncDefaultRetryAfter cNetworking) $
+                    fmap toSeconds mbCurrentRetryAfter
+
+              let moveProgressOnFixable = composeFuncList
+                    [ whenFunc (rcTotalRetries == 0) incProgressFixableErrors
+                    , whenFunc (rcTotalRetries == 0) incProgress
+                    ]
```

Comment:
And a couple of more things: I suppose, in all three branches, this can be composed in a more compact way. E.g. in the first case it is just `if .. then .. else`.

Also, this code will be more readable if we used consistently either `errorsOccured` or `rcTotalRetries == 0` in all the `case` branches.

Reply:
Thanks, that's exactly what I meant.